File: java/0206.0_Reverse_Linked_List.java
/*
iteration

执行用时：0 ms, 在所有 Java 提交中击败了100.00% 的用户
内存消耗：41.1 MB, 在所有 Java 提交中击败了24.23% 的用户
通过测试用例：28 / 28
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode another = null, node = head;
        while (node != null) {
            ListNode next_part = node.next;
            node.next = another;
            another = node;
            // move to next part
            node = next_part;
        }
        return another;
    }
}


/*
recursion

执行用时：0 ms, 在所有 Java 提交中击败了100.00% 的用户
内存消耗：41.3 MB, 在所有 Java 提交中击败了8.97% 的用户
通过测试用例：28 / 28
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode tail = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return tail;
    }
}

File: cpp/0206.0_Reverse_Linked_List.cpp
/*
iteration

执行用时：4 ms, 在所有 C++ 提交中击败了94.50% 的用户
内存消耗：8.1 MB, 在所有 C++ 提交中击败了53.96% 的用户
通过测试用例：28 / 28
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* another = nullptr, *node = head;
        while (node) {
            ListNode* next_part = node->next;
            node->next = another;
            another = node;
            node = next_part;
        }
        return another;
    }
};


/*
recursion

执行用时：4 ms, 在所有 C++ 提交中击败了94.50% 的用户
内存消耗：8.3 MB, 在所有 C++ 提交中击败了6.89% 的用户
通过测试用例：28 / 28
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head or !head->next) {
            return head;
        }
        ListNode* tail = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return tail;
    }
};

